blog/2021-05-12-build-an-ecommerce-app-with-temporal-part-1.md

```diff
@@ -0,0 +1,271 @@
+import TabItem from '@theme/TabItem';
+
+<!--truncate-->
+
+## Introduction
+
+Temporal promises to help you build invincible apps.
+To make this possible, new design patterns are introduced that are very different from the ones used in a traditional web app architecture.
+
+Instead of letting your API endpoints talk to a database over the network, they would instead call in-memory _Workflows_ that store state internally.
+Temporal handles persisting the state of your Workflows and distributes your Workflow between Workers as necessary.
+You, as the developer are responsible for implementing [Workflows](https://docs.temporal.io/docs/go/workflows) and [Activities](https://docs.temporal.io/docs/go/activities) as normal Go code.
+Meanwhile, Temporal handles the data persistence and horizontal scaling for you.
+
+In this blog post, I'll demonstrate how to build a shopping cart using long-living Workflows.
+You can find the [full source code for this shopping cart on GitHub](https://github.com/vkarpov15/temporal-ecommerce).
+
+## Shopping cart Workflow
+
+In a traditional web app architecture, a user's shopping cart is stored as a row or document in a database.
+While you can store shopping carts in a separate database using Temporal, you have another option: you can represent a shopping cart as a long-living Workflow.
+
+A Workflow is a Go function that takes 2 parameters: a Temporal Workflow context `ctx` and an arbitrary `value`. 
+It can run for an arbitrarily long period of time, as Temporal can handle pausing and restarting the Workflow.
+Lastly, it is able to share its state via _queries_ and modify its state in response to _signals_.
+
+The following is a simplified shopping cart that adds a new product to the cart every time it receives an `updateCart` signal.
+
+```go
+package app
+
+import (
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	CartItem struct {
+		ProductId int
+		Quantity  int
+	}
+
+	CartState struct {
+		Items []CartItem
+		Email string
+	}
+)
+
+func CartWorkflowExample(ctx workflow.Context, state CartState) error {
+	logger := workflow.GetLogger(ctx)
+
+	err := workflow.SetQueryHandler(ctx, "getCart", func(input []byte) (CartState, error) {
+		return state, nil
+	})
+	if err != nil {
+		logger.Info("SetQueryHandler failed.", "Error", err)
+		return err
+	}
+
+	channel := workflow.GetSignalChannel(ctx, "cartMessages")
+	selector := workflow.NewSelector(ctx)
+
+	selector.AddReceive(channel, func(c workflow.ReceiveChannel, _ bool) {
+		var signal interface{}
+		c.Receive(ctx, &signal)
+		state.Items = append(state.Items, CartItem{ProductId: 0, Quantity: 1})
+	})
+
+	for {
+		selector.Select(ctx)
+	}
+
+	return nil
+}
+```
+
+To run a Workflow, you need to create a Worker process.
+A Temporal _Worker_ listens for events on a queue and has a list of registered Workflows that it can run in response to messages on the queue.
+Below is the largely-boilerplate `worker/main.go` file:
+
+<!--SNIPSTART temporal-ecommerce-worker-->
```

Comment:
If you have these placeholders here, then no need to check in the code between the placeholders.

The snippet gets merged in at build time.